projects/power_selection/test/test_power_selection.c

```diff
@@ -0,0 +1,177 @@
+#include "adc.h"
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "log.h"
+#include "ms_test_helpers.h"
+#include "power_selection.h"
+#include "power_selection_events.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+
+#define TEST_CAN_DEVICE_ID 0x1
+
+static CanStorage s_can_storage = { 0 };
+static CanSettings s_can_settings = {
+  .device_id = TEST_CAN_DEVICE_ID,
+  .bitrate = CAN_HW_BITRATE_500KBPS,
+  .rx_event = POWER_SELECTION_CAN_EVENT_RX,
+  .tx_event = POWER_SELECTION_CAN_EVENT_TX,
+  .fault_event = POWER_SELECTION_CAN_EVENT_FAULT,
+  .tx = { GPIO_PORT_A, 12 },
+  .rx = { GPIO_PORT_A, 11 },
+  .loopback = true,
+};
+
+static GpioAddress s_aux_status_addresses[3] = {
+  { .port = GPIO_PORT_A, .pin = 0 },  // aux voltage sense
+  { .port = GPIO_PORT_A, .pin = 3 },  // aux temp sense
+  { .port = GPIO_PORT_A, .pin = 5 }   // aux current sense  //not needed in the module currently
+};
+static GpioAddress s_dcdc_address = { .port = GPIO_PORT_A, .pin = 9 };
+static AdcChannel s_aux_channels[3] = { ADC_CHANNEL_0, ADC_CHANNEL_1, ADC_CHANNEL_2 };
+
+static uint16_t s_aux_volt_value = 0;
+static uint16_t s_aux_temp_value = 0;
+StatusCode TEST_MOCK(adc_read_raw)(AdcChannel adc_channel, uint16_t *reading) {
+  if (adc_channel == s_aux_channels[0]) {
+    *reading = s_aux_volt_value;
+  } else if (adc_channel == s_aux_channels[1]) {
+    *reading = s_aux_temp_value;
+  }
+  return STATUS_CODE_OK;
+}
+
+int counter = 0;
+uint16_t aux_volt = 0;
+uint16_t aux_temp = 0;
+uint16_t dcdc_status = 0;
+StatusCode prv_test_power_selection_callback_handler(const CanMessage *msg, void *context,
+                                                     CanAckStatus *ack_reply) {
+  TEST_ASSERT_EQUAL(SYSTEM_CAN_MESSAGE_AUX_BATTERY_STATUS, msg->msg_id);
+  counter++;
+  aux_volt = msg->data_u16[0];
+  aux_temp = msg->data_u16[1];
+  dcdc_status = msg->data_u16[2];
+  return STATUS_CODE_OK;
+}
+
+static bool supposed_to_fail = false;
+static StatusCode prv_can_simple_ack(CanMessageId msg_id, uint16_t device, CanAckStatus status,
+                                     uint16_t num_remaining, void *context) {
+  if (supposed_to_fail) {
+    TEST_ASSERT_EQUAL(status, CAN_ACK_STATUS_INVALID);
```

Comment:
in future you should probably do something like
```
TEST_ASSERT_NOT_EQUAL(status, CAN_ACK_STATUS_OK);
```
because there could be other ack failures, not just `CAN_ACK_STATUS_INVALID`. This is fine here though.

Reply:
i'll do that